utils/fixed_decimal/src/decimal.rs

```diff
@@ -618,6 +618,92 @@ impl FixedDecimal {
         self.check_invariants();
     }
 
+    /// Concatenate another `FixedDecimal` into this `FixedDecimal`.
+    ///
+    /// The two decimals must represent non-overlapping ranges of magnitudes. If the range of
+    /// nonzero digits overlaps between `self` and `other`, an `Err` is returned.
+    ///
+    /// The magnitude range of `self` will be increased if `other` covers a larger range.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fixed_decimal::FixedDecimal;
+    ///
+    /// let mut integer = FixedDecimal::from(123);
+    /// let fraction = FixedDecimal::from(456)
+    ///     .multiplied_pow10(-3)
+    ///     .expect("in-range");
+    ///
+    /// assert_eq!("123.456", integer.concatenated(fraction).expect("nonoverlapping").to_string());
+    /// ```
+    pub fn concatenated(mut self, other: FixedDecimal) -> Result<Self, Error> {
+        match self.concatenate(other) {
+            Ok(()) => Ok(self),
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Concatenate another `FixedDecimal` into this `FixedDecimal`.
+    ///
+    /// The two decimals must represent non-overlapping ranges of magnitudes. If the range of
+    /// nonzero digits overlaps between `self` and `other`, an `Err` is returned.
+    ///
+    /// The magnitude range of `self` will be increased if `other` covers a larger range.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fixed_decimal::FixedDecimal;
+    ///
+    /// let mut integer = FixedDecimal::from(123);
+    /// let fraction = FixedDecimal::from(456)
+    ///     .multiplied_pow10(-3)
+    ///     .expect("nonoverlapping");
+    ///
+    /// integer.concatenate(fraction);
+    ///
+    /// assert_eq!("123.456", integer.to_string());
+    /// ```
+    pub fn concatenate(&mut self, other: FixedDecimal) -> Result<(), Error> {
+        let self_left = self.nonzero_magnitude_left();
+        let self_right = self.nonzero_magnitude_right();
+        let other_left = other.nonzero_magnitude_left();
+        let other_right = other.nonzero_magnitude_right();
+        if self.is_zero() {
+            // Operation will succeed. We can move the digits into self.
+            self.digits = other.digits;
+            self.magnitude = other.magnitude;
+        } else if other.is_zero() {
+            // No changes to the digits are necessary.
```

Comment:
nit: can we just return?

Reply:
No, because the upper-lower range might still differ.

Reply:
ah